File: src/main/java/org/simbrain/util/piccolo/TiledData.java
package org.simbrain.util.piccolo;

import com.Ostermiller.util.CSVParser;
import com.thoughtworks.xstream.annotations.XStreamAlias;
import com.thoughtworks.xstream.annotations.XStreamConverter;
import com.thoughtworks.xstream.converters.extended.ToAttributedValueConverter;

import java.io.*;
import java.util.ArrayList;
import java.util.Base64;
import java.util.List;
import java.util.zip.GZIPInputStream;
import java.util.zip.InflaterInputStream;

/**
 * Tiled {@code <data>} tag object representation.
 *
 * Cannot handle xml encoded data for now.
 */
@XStreamAlias("data")
@XStreamConverter(value = ToAttributedValueConverter.class, strings = {"data"})
public class TiledData {

    /**
     * The encoding used to encode the tile layer data. When used, it can be “base64” and “csv” at the moment.
     */
    private String encoding;

    /**
     * The compression used to compress the tile layer data. Tiled supports “gzip” and “zlib”.
     */
    private String compression;

    /**
     * Raw content of the data.
     *
     * Not handling embedded image for now.
     */
    private char[] data;

    /**
     * The flat list of tile id from the raw data.
     */
    private transient List<Integer> gid = null;

    /**
     * Get the list of tile id.
     * Result is computed from raw data and stored in field on the first call.
     * After the first call this just returns the values stored in field.
     *
     * @return the list of tile id this data represents.
     */
    public List<Integer> getGid() {
        if (gid == null) {
            gid = new ArrayList<>();
            if ("csv".equals(encoding)) {
                CSVParser parser = new CSVParser(new CharArrayReader(data));
                String[][] allValues;
                try {
                    allValues = parser.getAllValues();
                } catch (IOException e) {
                    e.printStackTrace();
                    return null;
                }

                for (String[] row : allValues) {
                    for (String cell : row) {
                        if (!cell.isEmpty()) { // trailing commas create extra empty elements. filter them out.
                            gid.add(Integer.parseInt(cell));
                        }
                    }
                }
            } else if ("base64".equals(encoding)) {

                // remove spaces and line breaks
                String tempData = String.valueOf(data).replaceAll("[ \n]", "");
                data = tempData.toCharArray();

                byte[] decoded = Base64.getDecoder().decode(String.valueOf(data));
                if ("gzip".equals(compression)) {
                    try (InputStream inputStream = new GZIPInputStream(new ByteArrayInputStream(decoded))) {
                        decoded = inputStream.readAllBytes();
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                } else if ("zlib".equals(compression)) {
                    InputStream inputStream = new InflaterInputStream(new ByteArrayInputStream(decoded));
                    try {
                        decoded = inputStream.readAllBytes();
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }
                for (int i = 0; i < decoded.length; i += 4) {
                    // interpreted as an array of unsigned 32-bit integers using little-endian byte ordering.
                    // see https://doc.mapeditor.org/en/stable/reference/tmx-map-format/#tmx-data
                    gid.add(decoded[i] | decoded[i + 1] << 8 | decoded[i + 2] << 16 | decoded[i + 3] << 24);
                }
            }
        }
        return gid;
    }

}

File: src/main/kotlin/org/simbrain/util/piccolo/TiledData.kt
package org.simbrain.util.piccolo

import com.Ostermiller.util.CSVParser
import com.thoughtworks.xstream.annotations.XStreamAlias
import com.thoughtworks.xstream.converters.MarshallingContext
import com.thoughtworks.xstream.converters.UnmarshallingContext
import com.thoughtworks.xstream.converters.reflection.ReflectionConverter
import com.thoughtworks.xstream.converters.reflection.ReflectionProvider
import com.thoughtworks.xstream.io.HierarchicalStreamReader
import com.thoughtworks.xstream.io.HierarchicalStreamWriter
import com.thoughtworks.xstream.mapper.Mapper
import org.simbrain.network.NetworkModel
import org.simbrain.network.core.Network
import java.io.ByteArrayInputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.*
import java.util.zip.GZIPInputStream
import java.util.zip.InflaterInputStream

/**
 * Tiled `<data>` tag object representation.
 *
 * Cannot handle xml encoded data for now.
 */
@XStreamAlias("data")
class TiledData(val gid: MutableList<MutableList<Int>>) {
    constructor(width: Int, height: Int) : this(MutableList(height) { MutableList(width) { 0 } })
    operator fun get(x: Int, y: Int) = gid[y][x]
    operator fun set(x: Int, y: Int, tileId: Int) {
        gid[y][x] = tileId
    }
}

/**
 * Custom serializer that stores [Network.networkModels], which is a map, as a flat list of [NetworkModel]s.
 */
class TiledDataConverter(mapper: Mapper, reflectionProvider: ReflectionProvider) :
    ReflectionConverter(mapper, reflectionProvider, TiledData::class.java) {

    override fun marshal(source: Any?, writer: HierarchicalStreamWriter, context: MarshallingContext) {
        val data = source as TiledData
        writer.addAttribute("encoding", "csv")
        val csv = data.gid.joinToString("\n") { it.joinToString(",") }
        writer.setValue(csv)
    }

    /**
     * Decode the raw [data] into a list of tile ids.
     *
     * @return the list of tile id this data represents.
     */
    private fun decodeData(value: String, encoding: String, compression: String?): List<List<Int>> {

        fun decodeCSV() =
            CSVParser(value.byteInputStream()).allValues
                .map { row ->
                    row.filter { it.isNotEmpty() }
                        .map { it.toInt() }
                }

        fun decodeBase64() = value.split("[ \n]".toRegex())
            .map { row -> row.let { Base64.getDecoder().decode(it)!! } }


        fun ByteArray.decompressGzip() = GZIPInputStream(ByteArrayInputStream(this))
            .use { it.readAllBytes()!! }

        fun ByteArray.decompressZlib() = InflaterInputStream(ByteArrayInputStream(this)).readAllBytes()!!

        fun ByteBuffer.asIntSequence() = sequence {
            while(hasRemaining()) {
                yield(int)
            }
        }

        return when (encoding) {
            "csv" -> decodeCSV()
            "base64" -> when(compression) {
                "gzip" -> decodeBase64().map { it.decompressGzip() }
                "zlib" -> decodeBase64().map { it.decompressZlib() }
                else -> decodeBase64()
            }.map { ByteBuffer.wrap(it).apply { order(ByteOrder.LITTLE_ENDIAN) }.asIntSequence().toList() }
            else -> throw IllegalStateException("Unknown encoding $encoding")
        }

    }

    override fun unmarshal(reader: HierarchicalStreamReader, context: UnmarshallingContext): Any {
        val encoding = reader.getAttribute("encoding")
        val compression = reader.getAttribute("compression")
        val gid = decodeData(reader.value, encoding, compression)
            .map { it.toMutableList() }
            .toMutableList()
        return TiledData(gid)
    }
}